rcl_action/src/rcl_action/action_server.c

```diff
@@ -0,0 +1,759 @@
+    goto cleanup;
+  }
+
+  // Allocate space in response
+  rcl_ret_t ret = rcl_action_cancel_response_init(
+    cancel_response, num_goals_to_cancel, allocator);
+  if (RCL_RET_OK != ret) {
+    if (RCL_RET_BAD_ALLOC == ret) {
+      ret_final = RCL_RET_BAD_ALLOC;  // error already set
+    }
+    ret_final = RCL_RET_ERROR;  // error already set
+    goto cleanup;
+  }
+
+  // Transition goals to canceling and add to response
+  rcl_action_goal_handle_t * goal_handle;
+  for (size_t i = 0; i < num_goals_to_cancel; ++i) {
+    goal_handle = goal_handles_to_cancel[i];
+    ret = rcl_action_update_goal_state(goal_handle, GOAL_EVENT_CANCEL);
+    if (RCL_RET_OK == ret) {
+      ret = rcl_action_goal_handle_get_info(
+        goal_handle, &cancel_response->msg.goals_canceling.data[i]);
+    }
+    if (RCL_RET_OK != ret) {
+      ret_final = RCL_RET_ERROR;  // error already set
+    }
+  }
+cleanup:
+  allocator.deallocate(goal_handles_to_cancel, allocator.state);
+  return ret_final;
+}
+
+rcl_ret_t
+rcl_action_send_cancel_response(
+  const rcl_action_server_t * action_server,
+  void * ros_cancel_response)
+{
+  SEND_SERVICE_RESPONSE(cancel);
+}
+
+const char *
+rcl_action_server_get_action_name(const rcl_action_server_t * action_server)
+{
+  if (!rcl_action_server_is_valid(action_server)) {
+    return NULL;  // error already set
+  }
+  return action_server->impl->action_name;
+}
+
+const rcl_action_server_options_t *
+rcl_action_server_get_options(const rcl_action_server_t * action_server)
+{
+  if (!rcl_action_server_is_valid(action_server)) {
+    return NULL;  // error already set
+  }
+  return &action_server->impl->options;
+}
+
+rcl_ret_t
+rcl_action_server_get_goal_handles(
+  const rcl_action_server_t * action_server,
+  rcl_action_goal_handle_t *** goal_handles,
+  size_t * num_goals)
+{
+  if (!rcl_action_server_is_valid(action_server)) {
+    return RCL_RET_ACTION_SERVER_INVALID;  // error already set
+  }
+  RCL_CHECK_ARGUMENT_FOR_NULL(goal_handles, RCL_RET_INVALID_ARGUMENT);
+  RCL_CHECK_ARGUMENT_FOR_NULL(num_goals, RCL_RET_INVALID_ARGUMENT);
+  *goal_handles = action_server->impl->goal_handles;
+  *num_goals = action_server->impl->num_goal_handles;
+  return RCL_RET_OK;
+}
+
+bool
+rcl_action_server_goal_exists(
+  const rcl_action_server_t * action_server,
+  const rcl_action_goal_info_t * goal_info)
+{
+  if (!rcl_action_server_is_valid(action_server)) {
+    return false;  // error already set
+  }
+  RCL_CHECK_ARGUMENT_FOR_NULL(goal_info, false);
+
+  rcl_action_goal_info_t gh_goal_info = rcl_action_get_zero_initialized_goal_info();
+  rcl_ret_t ret;
+  for (size_t i = 0; i < action_server->impl->num_goal_handles; ++i) {
+    ret = rcl_action_goal_handle_get_info(action_server->impl->goal_handles[i], &gh_goal_info);
+    if (RCL_RET_OK != ret) {
+      RCL_SET_ERROR_MSG("failed to get info for goal handle");
+      return false;
+    }
+    // Compare UUIDs
+    if (uuidcmp(gh_goal_info.uuid, goal_info->uuid)) {
```

Comment:
nit could return this directly

Reply:
I don't think so. If it is false, there still might be a match in a future iteration of the for-loop.

Reply:
Ooops